From 03821afe0754ef7083a28406e60fb69ff4e2c89a Mon Sep 17 00:00:00 2001
From: Shwetha G K <kshwetha@qti.qualcomm.com>
Date: Mon, 23 Jun 2025 17:22:03 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL: wifi: mac80211: Add support to notify advertised TTLM
 status

Introduce infrastructure in mac80211 to notify userspace about
TID-to-Link Mapping (TTLM) status updates for AP interfaces
operating in drivers with offload support. Extend ieee80211_link_data_ap
to store TTLM status indications received from the driver and
schedule a workqueue to propagate these updates to userspace applications.

Report the following TTLM status events:
- Mapping Switch Time (MST) in TSF units, converted from TU by lower
  layers.
- MST expiry, indicating when the new mapping becomes active on TIDs.
- Expected Duration (ED) expiry, signaling the end of the advertised
  mapping and reversion to the default mapping where all TIDs are mapped
  to all valid links of the AP MLD.

Signed-off-by: Shwetha G K <kshwetha@qti.qualcomm.com>
Signed-off-by: Amith A <amitajit@qti.qualcomm.com>
---
 include/net/mac80211.h     | 16 +++++++++++
 net/mac80211/cfg.c         | 54 ++++++++++++++++++++++++++++++++++++++
 net/mac80211/ieee80211_i.h |  8 ++++++
 net/mac80211/link.c        |  2 ++
 4 files changed, 80 insertions(+)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index f9ff264..80020cc 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -8366,4 +8366,20 @@ int ieee80211_6ghz_power_mode_change(struct wiphy *wiphy,
 void ieee80211_link_removal_count_update(struct ieee80211_vif *vif,
 					 unsigned int link_id, u16 count);
 
+/**
+ * ieee80211_advertised_ttlm_evt_notify - notify about driver offloaded
+ *	advertised TTLM status
+ *
+ * @vif: interface
+ * @mst_tsf: Mapping Switch Time in TSF if status is indicating updated MST in
+ *	TSF (ADVERTISED_TTLM_SWITCH_TIMER_TSF)
+ * @staus: Status of the advertised TTLM in offloaded mode.
+ * @link_id: link id of the interface on which status received.
+ */
+void
+ieee80211_advertised_ttlm_evt_notify(struct ieee80211_vif *vif,
+				     u16 mst_tsf,
+				     enum advertised_ttlm_status_type status,
+				     unsigned int link_id);
+
 #endif /* MAC80211_H */
diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index a3382f1..fce6ed7 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -209,6 +209,23 @@ static int ieee80211_set_ap_mbssid_options(struct ieee80211_sub_if_data *sdata,
 	return 0;
 }
 
+void ieee80211_advertised_ttlm_evt_notify_work(struct wiphy *wiphy,
+					       struct wiphy_work *work)
+{
+	struct ieee80211_link_data *link =
+		container_of(work, struct ieee80211_link_data,
+			     u.ap.advertised_ttlm_evt_notify_work);
+	struct ieee80211_sub_if_data *sdata = link->sdata;
+
+	if (!ieee80211_sdata_running(sdata))
+		return;
+
+	cfg80211_adv_ttlm_evt_notify(sdata->dev, GFP_KERNEL,
+				     link->u.ap.advertised_ttlm_status,
+				     link->u.ap.advertised_ttlm_mst_tsf,
+				     link->link_id);
+}
+
 static struct wireless_dev *ieee80211_add_iface(struct wiphy *wiphy,
 						const char *name,
 						unsigned char name_assign_type,
@@ -5652,6 +5669,43 @@ ieee80211_obss_color_collision_notify(struct ieee80211_vif *vif,
 }
 EXPORT_SYMBOL_GPL(ieee80211_obss_color_collision_notify);
 
+void
+ieee80211_advertised_ttlm_evt_notify(struct ieee80211_vif *vif,
+				     u16 mst_tsf,
+				     enum advertised_ttlm_status_type status,
+				     unsigned int link_id)
+{
+	struct ieee80211_sub_if_data *sdata = vif_to_sdata(vif);
+	struct wireless_dev *wdev = ieee80211_vif_to_wdev(vif);
+	struct ieee80211_link_data *link;
+
+	if (sdata->vif.type != NL80211_IFTYPE_AP)
+		return;
+
+	if (!wdev)
+		return;
+
+	if (WARN_ON(link_id >= IEEE80211_MLD_MAX_NUM_LINKS))
+		return;
+
+	rcu_read_lock();
+
+	link = rcu_dereference(sdata->link[link_id]);
+	if (WARN_ON(!link)) {
+		rcu_read_unlock();
+		return;
+	}
+
+	link->u.ap.advertised_ttlm_status = status;
+	link->u.ap.advertised_ttlm_mst_tsf = mst_tsf;
+
+	wiphy_work_queue(sdata->local->hw.wiphy,
+			 &link->u.ap.advertised_ttlm_evt_notify_work);
+
+	rcu_read_unlock();
+}
+EXPORT_SYMBOL_GPL(ieee80211_advertised_ttlm_evt_notify);
+
 static int
 ieee80211_color_change(struct wiphy *wiphy, struct net_device *dev,
 		       struct cfg80211_color_change_settings *params,
diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index 14deaff..a126380 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -1061,6 +1061,11 @@ struct ieee80211_link_data_ap {
 	/* to be used after channel switch. */
 	struct cfg80211_beacon_data *next_beacon;
 	bool after_beacon_cu;
+
+	/* advertised ttlm event info */
+	u16 advertised_ttlm_mst_tsf;
+	enum advertised_ttlm_status_type advertised_ttlm_status;
+	struct wiphy_work advertised_ttlm_evt_notify_work;
 };
 
 struct ieee80211_link_data {
@@ -2127,6 +2132,9 @@ void ieee80211_color_change_finalize_work(struct wiphy *wiphy,
 void ieee80211_color_collision_detection_work(struct wiphy *wiphy,
 					      struct wiphy_work *work);
 
+/* advertised ttlm handling */
+void ieee80211_advertised_ttlm_evt_notify_work(struct wiphy *wiphy,
+					       struct wiphy_work *work);
 /* interface handling */
 #define MAC80211_SUPPORTED_FEATURES_TX	(NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM | \
 					 NETIF_F_HW_CSUM | NETIF_F_SG | \
diff --git a/net/mac80211/link.c b/net/mac80211/link.c
index c24233c..f1cb54d 100644
--- a/net/mac80211/link.c
+++ b/net/mac80211/link.c
@@ -141,6 +141,8 @@ skip_vlan_conf:
 			ieee80211_color_change_finalize_work);
 	wiphy_delayed_work_init(&link->color_collision_detect_work,
 				ieee80211_color_collision_detection_work);
+	wiphy_work_init(&link->u.ap.advertised_ttlm_evt_notify_work,
+			ieee80211_advertised_ttlm_evt_notify_work);
 	INIT_LIST_HEAD(&link->assigned_chanctx_list);
 	INIT_LIST_HEAD(&link->reserved_chanctx_list);
 	wiphy_work_init(&link->dfs_cac_timer_work,
-- 
2.34.1

